Use `e0` as a separate field together with `x0::NTuple` in `BirchMurnaghan`

In #26 I said I wanted to deprecate `NTuple` as `x0` in `BirchMurnaghan{N,T}`, but there
is still 1 possibility left:

```julia
struct BirchMurnaghan{N,T} <: FiniteStrainEossParameters{N,T}
    x0::NTuple{N,T}
    e0::T
end
```

Let's see how far can we go.

<!-- Edit the body of your new issue then click the ✓ "Create Issue" button in the top right of the editor. The first line will be the issue title. Assignees and Labels follow after a blank line. Leave an empty line before beginning the body of the issue. -->

This brings 1 issue: the creation of types with units.

The [current implementation](https://github.com/MineralsCloud/EquationsOfStateOfSolids.jl/blob/2a7a90d311db15ba44056e256749cdfb7c00ebdf/src/Collections.jl#L30-L33) cannot create a `BirchMurnaghan` with units:

```julia
julia> x = BirchMurnaghan([1u"m^3",2u"GPa",3.0,4u"eV"])
ERROR: AssertionError: 2 <= N <= 5
Stacktrace:
 [1] BirchMurnaghan{1,Tuple{Quantity{Float64,𝐋³,Unitful.FreeUnits{(m³,),𝐋³,nothing}},Quantity{Float64,𝐌 𝐋⁻¹ 𝐓⁻²,Unitful.FreeUnits{(GPa,),𝐌 𝐋⁻¹ 𝐓⁻²,nothing}},Quantity{Float64,NoDims,Unitful.FreeUnits{(),NoDims,nothing}}}}(::Tuple{Tuple{Quantity{Float64,𝐋³,Unitful.FreeUnits{(m³,),𝐋³,nothing}},Quantity{Float64,𝐌 𝐋⁻¹ 𝐓⁻²,Unitful.FreeUnits{(GPa,),𝐌 𝐋⁻¹ 𝐓⁻²,nothing}},Quantity{Float64,NoDims,Unitful.FreeUnits{(),NoDims,nothing}}}}, ::Quantity{Float64,𝐋² 𝐌 𝐓⁻²,Unitful.FreeUnits{(eV,),𝐋² 𝐌 𝐓⁻²,nothing}}) at /Users/qz/.julia/dev/EquationsOfStateOfSolids/src/Collections.jl:26
 [2] BirchMurnaghan(::Tuple{Tuple{Quantity{Float64,𝐋³,Unitful.FreeUnits{(m³,),𝐋³,nothing}},Quantity{Float64,𝐌 𝐋⁻¹ 𝐓⁻²,Unitful.FreeUnits{(GPa,),𝐌 𝐋⁻¹ 𝐓⁻²,nothing}},Quantity{Float64,NoDims,Unitful.FreeUnits{(),NoDims,nothing}}}}, ::Quantity{Float64,𝐋² 𝐌 𝐓⁻²,Unitful.FreeUnits{(eV,),𝐋² 𝐌 𝐓⁻²,nothing}}) at /Users/qz/.julia/dev/EquationsOfStateOfSolids/src/Collections.jl:30
 [3] BirchMurnaghan(::Array{Any,1}) at /Users/qz/.julia/dev/EquationsOfStateOfSolids/src/Collections.jl:32
 [4] BirchMurnaghan(::Tuple{Quantity{Float64,𝐋³,Unitful.FreeUnits{(m³,),𝐋³,nothing}},Quantity{Float64,𝐌 𝐋⁻¹ 𝐓⁻²,Unitful.FreeUnits{(GPa,),𝐌 𝐋⁻¹ 𝐓⁻²,nothing}},Quantity{Float64,NoDims,Unitful.FreeUnits{(),NoDims,nothing}}}, ::Vararg{Any,N} where N) at /Users/qz/.julia/dev/EquationsOfStateOfSolids/src/Collections.jl:34
 [5] BirchMurnaghan(::Array{Quantity{Float64,D,U} where U where D,1}) at /Users/qz/.julia/dev/EquationsOfStateOfSolids/src/Collections.jl:32
 [6] top-level scope at REPL[5]:1
 [7] include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1088
```

The reason is that `x0` and `e0` have different `eltype`s:

```julia
julia> [1u"m^3",2u"GPa",3.0] |> eltype
Quantity{Float64,D,U} where U where D

julia> 4u"eV" |> eltype
Quantity{Int64,𝐋² 𝐌 𝐓⁻²,Unitful.FreeUnits{(eV,),𝐋² 𝐌 𝐓⁻²,nothing}}
```

--- a/Project.toml
+++ b/Project.toml
@@ -7,6 +7,7 @@ version = "0.1.0"
 AutoHashEquals = "15f4f7f2-30c1-5605-9d31-71845cf9641f"
 ConstructionBase = "187b0558-2788-49d3-abe0-74a17ed4e7c9"
 LsqFit = "2fda8390-95c7-5789-9bda-21331edee243"
+UnPack = "3a884ed6-31ef-47d7-9d2a-63182c4928ed"
 Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
 
 [compat]

--- a/src/Collections.jl
+++ b/src/Collections.jl
@@ -75,29 +75,6 @@ volume_from_strain(::Lagrangian, v0) = f -> v0 * (2f + 1)^(3 / 2)
 volume_from_strain(::Natural, v0) = f -> v0 * exp(3f)
 volume_from_strain(::Infinitesimal, v0) = f -> v0 / (1 - f)^3
 
-fieldvalues(x::EossParameters) = x.x0
-
-Base.propertynames(::FiniteStrainEossParameters{2}) = (:v0, :b0, :e0)
-Base.propertynames(::FiniteStrainEossParameters{3}) = (:v0, :b0, :b′0, :e0)
-Base.propertynames(::FiniteStrainEossParameters{4}) = (:v0, :b0, :b′0, :b′′0, :e0)
-Base.propertynames(::FiniteStrainEossParameters{5}) = (:v0, :b0, :b′0, :b′′0, :b′′′0, :e0)
-
-function Base.getproperty(value::FiniteStrainEossParameters, name::Symbol)
-    if name == :v0
-        return value.x0[1]
-    elseif name == :b0
-        return value.x0[2]
-    elseif name == :b′0
-        return value.x0[3]
-    elseif name == :b′′0
-        return value.x0[4]
-    elseif name == :b′′′0
-        return value.x0[5]
-    else
-        return getfield(value, name)
-    end
-end
-
 function Base.show(io::IO, eos::EossParameters)  # Ref: https://github.com/mauro3/Parameters.jl/blob/3c1d72b/src/Parameters.jl#L542-L549
     if get(io, :compact, false)
         Base.show_default(IOContext(io, :limit => true), eos)

--- a/src/Collections.jl
+++ b/src/Collections.jl
@@ -1,7 +1,7 @@
 module Collections
 
 using AutoHashEquals: @auto_hash_equals
-using Unitful: AbstractQuantity, @u_str
+using UnPack: @unpack
 
 export BirchMurnaghan3rd,
     Eulerian,
@@ -11,6 +11,7 @@ export BirchMurnaghan3rd,
     energyeos,
     pressureeos,
     bulkmoduluseos,
+    orderof,
     nextorder,
     strain_from_volume,
     volume_from_strain
@@ -44,15 +45,15 @@ energyeos(p::EossParameters) = EnergyEoss(p)
 pressureeos(p::EossParameters) = PressureEoss(p)
 bulkmoduluseos(p::EossParameters) = BulkModulusEoss(p)
 
-function (eos::EnergyEoss{<:BirchMurnaghan{3}})(v, e0 = zero(eos.params.x0[1] * eos.params.x0[3]))
-    v0, b0, b′0 = eos.params.x0
+function (eos::EnergyEoss{<:BirchMurnaghan3rd})(v)
+    @unpack v0, b0, b′0, e0 = eos.params
     x = cbrt(v0 / v)
     y = x^2 - 1
     return 9 / 16 * b0 * v0 * y^2 * (6 - 4 * x^2 + b′0 * y) + e0
 end
 
-function (eos::PressureEoss{<:BirchMurnaghan{3}})(v)
-    v0, b0, b′0 = eos.params.x0
+function (eos::PressureEoss{<:BirchMurnaghan3rd})(v)
+    @unpack v0, b0, b′0 = eos.params
     f = strain_from_volume(Eulerian(), v0)(v)
     return 3f / 2 * b0 * sqrt(2f + 1)^5 * (2 + 3f * (b′0 - 4))
 end

--- a/src/Collections.jl
+++ b/src/Collections.jl
@@ -3,7 +3,7 @@ module Collections
 using AutoHashEquals: @auto_hash_equals
 using Unitful: AbstractQuantity, @u_str
 
-export BirchMurnaghan,
+export BirchMurnaghan3rd,
     Eulerian,
     Lagrangian,
     Natural,
@@ -19,19 +19,15 @@ abstract type EossParameters{T} end
 
 abstract type FiniteStrainEossParameters{N,T} <: EossParameters{T} end
 
-struct BirchMurnaghan{N,T} <: FiniteStrainEossParameters{N,T}
-    x0::NTuple{N,T}
+struct BirchMurnaghan3rd{T} <: FiniteStrainEossParameters{3,T}
+    v0::T
+    b0::T
+    b′0::T
     e0::T
-    function BirchMurnaghan{N,T}(x0, e0) where {N,T}
-        @assert 2 <= N <= 5
-        new(x0, e0)
-    end
+    BirchMurnaghan3rd{T}(v0, b0, b′0, e0 = zero(v0 * b0)) where {T} = new(v0, b0, b′0, e0)
 end
-BirchMurnaghan(x0::NTuple{N,T}, e0 = zero(x0[1] * x0[2])) where {N,T} =
-    BirchMurnaghan{N,T}(x0, e0)
-BirchMurnaghan(arr::AbstractArray{T}) where {T} =
-    BirchMurnaghan(Tuple(arr[1:end-1]), arr[end])
-BirchMurnaghan(args...) = BirchMurnaghan([args...])
+BirchMurnaghan3rd(arr::AbstractArray) = BirchMurnaghan3rd{eltype(arr)}(arr...)
+BirchMurnaghan3rd(args...) = BirchMurnaghan3rd([args...])
 
 abstract type EquationOfStateOfSolids{T<:EossParameters} end
 struct EnergyEoss{T} <: EquationOfStateOfSolids{T}

--- a/src/Collections.jl
+++ b/src/Collections.jl
@@ -58,7 +58,7 @@ function (eos::PressureEoss{<:BirchMurnaghan3rd})(v)
     return 3f / 2 * b0 * sqrt(2f + 1)^5 * (2 + 3f * (b′0 - 4))
 end
 
-nextorder(::Type{BirchMurnaghan{N}}) where {N} = BirchMurnaghan{N + 1}
+orderof(::FiniteStrainEossParameters{N}) where {N} = N
 
 abstract type FiniteStrain end  # Trait
 struct Eulerian <: FiniteStrain end

